Post/methods arguments 2 (#413)
ombulabs/blog: _posts/2020-03-15-method-s-arguments-pt-1.markdown

---
layout: post
title: Ruby Method's Arguments
date: 2020-03-15 09:30:00
categories: [ruby, learning]
author: arieljuod
---

> Ruby is an object oriented language where everything is an object (even methods are objects of the class Method!), so everything we need to do is done by calling methods on objects. That also means that methods have to provide a lot of flexibility because they are used everywhere.

Reviewer: @arielj maybe add a link to the method class?

Author: I had part1 duplicated here, I removed the file. This PR is only for part2 but I'll update this suggestions for part1 on another PR

> Ruby provides a lot of options to pass arguments to our methods, so we'll make this topic a series so it's not too long. We'll split the options in different categories and then break down everything with some examples and/or use cases.

Reviewer 2: ```suggestion
Ruby provides a lot of options to pass arguments to our methods, so we'll make this topic a series so it's not too long. We'll split the options into different categories and then break down everything with some examples and/or use cases.
```

Reviewer 2: Maybe you should say something here about this being the second part of the series. I know it's in the title, but because this is the same opening as the previous blog, it might be sightly confusing.

> But first, we need to define the glossary:

Reviewer 2: I think you can remove this as you did in the last post.

<!--more-->

# Methods Vs. Functions

> Both methods and functions serve the same purpose: encapsulate a piece of code to reuse it with ease. And some times we use them as sinonyms, but there's a conceptual difference:

Reviewer 2: ```suggestion
Both methods and functions serve the same purpose: encapsulate a piece of code to reuse it with ease. And some times we use them as synonyms, but there's a conceptual difference:
```

> Both methods and functions serve the same purpose: encapsulate a piece of code to reuse it with ease. And some times we use them as sinonyms, but there's a conceptual difference:

Reviewer: ```suggestion
Both methods and functions serve the same purpose: encapsulate a piece of code to reuse it with ease. And sometimes we use them as synonyms, but there's a conceptual difference:
```

- Functions return a result based only on the input, functions shouldn't depend on or modify the state of something that's not in the given inputs.
- Methods are always executed in the context of an object, so we can call a method with no arguments but we always have access to the method's state and we can modify it.

In Ruby, where everything is an object, we can never have real functions but we can have methods that won't use or modify the object's state if we want to have something similar.

> # Arguments Vs. Parameters

Reviewer: ```suggestion
# Arguments vs. Parameters
```

Another thing that's worth differentiating is the concept of arguments and parameters. When we define a method, the parameters are part of the definition, the parameters are the variables that will contain the values that we use when calling that method. Arguments are the actual values that we use when we call a method. So, in this code:

```ruby
def my_method(foo)
  # foo is a parameter when defining the method
end

my_method("bar")
# "bar" is the argument we use when calling the method
```

Most of the times I'll use the word "arguments" for simplicity, since a method defines the parameters and accepts arguments.

> # No arguments

Reviewer: ```suggestion
# No Arguments
```

We can have methods that have no input, like this:

```ruby
def pi # parenthesis are optional and the keyword "return" is optional too
  3.14
end
```

But given that this is a method, it has access to the object's state to do different things

```ruby
class Circle
  def initialize(radius)
    @radius = radius
  end

  def pi
    3.14
  end

  def area
    pi * @radius ** 2
  end
end

my_circle = Circle.new(3)
puts my_circle.area
# => 28.26

other_circle = Circle.new(4)
puts other_circle.area
# => 50.24
```

> # Positional arguments

Reviewer: ```suggestion
# Positional Arguments
```

This type of argument gets that name because the order you use matters. You can have 0 or more positional arguments, and you can have required and optional arguments.

> # Required positional arguments

Reviewer: ```suggestion
# Required Positional Arguments
```

These positional arguments are required when calling a method, it's not that it's required to use them when defining the method (you can have no arguments, only optionals, etc). If the method defines required parameters you have to provide an argument for each position.

```ruby
def foo(arg1)
  puts "arg1 is: #{arg1.inspect}"
end

foo(1)
# => arg1 is: 1

foo
# ArgumentError (wrong number of arguments (given 0, expected 1))
```

We can have as many as we want:

```ruby
def foo(arg1, arg2, arg3, arg4)
# and more and more...
```

> Note that methods with too many arguments are a sign of bad design since the method is probably doing too many things!

> # Optional positional arguments

Reviewer: ```suggestion
# Optional Positional Arguments
```

Sometimes we may want to allow the user to provide more input but not require them to do so. For that, we can use optional arguments for which we define a default value in case the user doesn't specify one. This is particularly useful when we want to have some default behavior for our method but give the user the option to modify that. If we just hardcoded the default value inside the method we wouldn't have this flexibility.

```ruby
def foo(arg1 = 'default value')
  puts "arg1 is: #{arg1.inspect}"
end

foo(1)
# => arg1 is: 1

foo
# => arg1 is: default value
```

We can have as many as we want, just like the required parameters.

> # Combining required and optional positional arguments

Reviewer: ```suggestion
# Combining Required and Optional Positional Arguments
```

Now we can combine both required and positional arguments in one method definition:

```ruby
def foo(arg1, arg2 = 'default value')
  puts "arg1 es: #{arg1.inspect}"
  puts "arg2 es: #{arg2.inspect}"
end

foo(1)
# => arg1 is: 1
#    arg2 is: default value

foo(1,2)
# => arg1 is: 1
#    arg2 is: 2
```

Usually, when we mix both types of arguments, we put the required arguments first and then the optional ones. But Ruby allows us to do weird things like:

```ruby
def foo(arg1, arg2 = 'default value', arg3)
  puts "arg1 is: #{arg1.inspect}"
  puts "arg2 is: #{arg2.inspect}"
  puts "arg3 is: #{arg3.inspect}"
end
```

But this can create confusing situations:

```ruby
foo(1, 2, 3)
# => arg1 is: 1
#    arg2 is: 2
#    arg3 is: 3

foo(1,2)
# => arg1 is: 1
#    arg2 is: default value
#    arg3 is: 2
```

> Notice that, on the second call, the second argument ends up at **arg3**. My first thought was that ruby first assigned the required arguments and then the optionals, but that's not the case. Check out this even more confusing example:

Reviewer: ```suggestion
Notice that, on the second call, the second argument ends up at **arg3**. My first thought was that Ruby first assigned the required arguments and then the optionals, but that's not the case. Check out this even more confusing example:
```

```ruby
def foo(arg1, arg2 = 'default value X', arg3 = 'default value Y', arg4)
  puts "arg1 is: #{arg1.inspect}"
  puts "arg2 is: #{arg2.inspect}"
  puts "arg3 is: #{arg3.inspect}"
  puts "arg4 is: #{arg4.inspect}"
end

foo(1, 2, 3, 4)
arg1 is: 1
arg2 is: 2
arg3 is: 3

> arg4 is: 4

Reviewer: @arielj It might display in a weird way because your code snippet is using `ruby` -- is there any way to improve syntax highlighting here? 

![Screen Shot 2020-04-08 at 10 23 54 AM](https://user-images.githubusercontent.com/17584/78795503-18aeb180-7983-11ea-91ef-2702237da236.png)

Author: yes, I forgot the # at te beginning like I used on the other snippets

foo(1, 2, 3)
arg1 is: 1
arg2 is: 2
arg3 is: "default value Y"
arg4 is: 3

foo(1, 2)
arg1 is: 1
arg2 is: "default value X"
arg3 is: "default value Y"
arg4 is: 2
```

What Ruby seems to do is:

> - assign leftmost arguments to the first parameters

Reviewer: ```suggestion
- Assign leftmost arguments to the first parameters
```

> - assign rightmost arguments to the last parameters

Reviewer: ```suggestion
- Assign rightmost arguments to the last parameters
```

> - if there are not enough arguments, raise an error

Reviewer: ```suggestion
- If there are not enough arguments, raise an error
```

> - if there are unused arguments, assign them to the optional arguments in order from left to right

Reviewer: ```suggestion
- If there are unused arguments, assign them to the optional arguments in order from left to right
```

> I would recommend avoiding this combination unless it's something with a really specific requirement since it's not too intuitive and it's hard to follow.

Having optional arguments not grouped will raise an error:

```ruby
def foo(arg1, arg2 = 'default value2', arg3, arg4 = 'default value4')
end

# SyntaxError ((irb):1: syntax error, unexpected '=', expecting ')')
```

> # Optional arguments based on other arguments

Reviewer: ```suggestion
# Optional Arguments Based on Other Arguments
```

foo(1)
# => arg1 is: 1
#    arg2 is: 2

foo(1,3)
# => arg1 is: 1
#    arg2 is: 3
```

This can even depend on the value of optional arguments:

```ruby
def foo(arg1 = 1, arg2 = arg1 * 2)
  puts "arg1 is: #{arg1.inspect}"
  puts "arg2 is: #{arg2.inspect}"
end

foo
# => arg1 is: 1
#    arg2 is: 2

foo(2)
# => arg1 is: 2
#    arg2 is: 4
```

The only requirement is that the arguments used in the default value were previously assigned, this won't work:

```ruby
def foo(arg1 = arg2 * 2, arg2 = 1)
  puts "arg1 is: #{arg1.inspect}"
  puts "arg2 is: #{arg2.inspect}"
end

foo
# NameError (undefined local variable or method `arg2' for main:Object)
```

Also, we don't need to have simple operations like that for the default value, we can assign a function that will be called when assigning the values:

```ruby
def two_times(x)
  x * 2
end

def foo(arg1 = 1, arg2 = two_times(arg1))
  puts "arg1 is: #{arg1.inspect}"
  puts "arg2 is: #{arg2.inspect}"
end

foo
# => arg1 is: 1
#    arg2 is: 2

foo(5)
# => arg1 is: 5
#    arg2 is: 10 # it's multiplying 5, not 1, it uses our value, not te default one

foo(5, 6)
# => arg1 is: 5
#    arg2 is: 6 # we are actually overriding that multiplication
```

> # Variable arguments

Reviewer: ```suggestion
# Variable Arguments
```
